### PR TITLE
added len of a block

### PR DESCRIPTION
**Pull request by DavideTisi**
add a __len__ function to the block.
I found it annoying not to be able to do len(block). 
It is not revolutionary so if you do not like you can also not merge it.

----
 📚 [Download documentation preview for this pull-request](https://nightly.link/lab-cosmo/metatensor/actions/artifacts/1558537230.zip)

### CODE REVIEW

**Comment by Luthaf** [comment]
Length of a block makes sense. As a related function, do we want to expose `block.shape` as well?

File: python/metatensor-core/metatensor/block.py [hunk 1]
```diff
@@ -182,6 +182,12 @@ def __reduce__(self):
             "TensorMap first"
         )
 
+    def __len__(self):
+        """
+        get len of the values array
```
Comment (Luthaf) [inline python/metatensor-core/metatensor/block.py:187]:
```suggestion
        Get the length of the values stored in this block (i.e. the number of samples in the block)
```

**Review by Luthaf — COMMENTED** [review]
we need the same function to be available for the TorchScript backend. This means adding the function in C++ `TensorBlockHolder` and registering it with TorchScript.

**Comment by DavideTisi** [comment]
> we need the same function to be available for the TorchScript backend. This means adding the function in C++ `TensorBlockHolder` and registering it with TorchScript.

ok but why this is not done for `__len__` in tensorMap ?

**Comment by Luthaf** [comment]
It is also done for `TensorMap.__len__`: https://github.com/lab-cosmo/metatensor/blob/bf8c0130961b61d9a0dd10c676cf8570f611cc57/metatensor-torch/src/register.cpp#L213

Although this is only exposed to TorchScript and not to C++. We could change this!

**Comment by DavideTisi** [comment]
yes i deleted my comment almost immediately when i saw that lines

**Comment by DavideTisi** [comment]
now it should be done

File: metatensor-torch/include/metatensor/torch/block.hpp [hunk 2]
```diff
@@ -59,11 +59,19 @@ class METATENSOR_TORCH_EXPORT TensorBlockHolder: public torch::CustomClassHolder
...
     }
 
-    /// get the len of the samples
+    /// Get the len of the samples
     int64_t len() const {
         return this->labels(0)->count();
     }
 
+    /// Get the shape of the values Tensor
+    torch::Tensor shape() const{
```
Comment (Luthaf) [inline metatensor-torch/include/metatensor/torch/block.hpp:68]: this should be returned as the same type as `torch::Tensor::sizes`, which I think is either `at::IntArrayRef` or `at::IntArray`.

File: metatensor-torch/include/metatensor/torch/block.hpp [hunk 3]
```diff
@@ -59,11 +59,19 @@ class METATENSOR_TORCH_EXPORT TensorBlockHolder: public torch::CustomClassHolder
         return this->labels(0);
     }
 
-    /// get the len of the samples
+    /// Get the len of the samples
```
Comment (Luthaf) [inline metatensor-torch/include/metatensor/torch/block.hpp:62]:
```suggestion
    /// Get the length of this block, i.e. the number of samples
```

File: python/metatensor-core/metatensor/block.py [hunk 4]
```diff
@@ -182,13 +182,22 @@ def __reduce__(self):
...
         (i.e. the number of samples in the block)
         """
         return len(self.values)
 
+    @property
+    def shape(self):
+        """
+        Get the shape of the values stored in this block
+        (i.e. the number of samples, of components, of properties
+        in the block)
```
Comment (Luthaf) [inline python/metatensor-core/metatensor/block.py:197]:
```suggestion
        Get the shape of the values  array in this block.
```

File: python/metatensor-core/tests/blocks.py [hunk 5]
```diff
@@ -41,6 +41,14 @@ def block_components():
     )
 
 
+def test_len(block):
```
Comment (Luthaf) [inline python/metatensor-core/tests/blocks.py:44]: could you also add the same tests in metatensor-torch?

File: metatensor-torch/include/metatensor/torch/block.hpp (hunk 2 above)
Comment (DavideTisi) [reply metatensor-torch/include/metatensor/torch/block.hpp:68]: which i cannot use, because it fails I think it is not TorchScriptable

**Comment by DavideTisi** [comment]
the shape code was wrong 
`at::IntArrayRef` is a reference so it need to reference to some variable that do not goes out of scope. 
so I create a shape variable for the class

it should be correct now, at least the test works

File: metatensor-torch/include/metatensor/torch/block.hpp (hunk 2 above)
Comment (Luthaf) [reply metatensor-torch/include/metatensor/torch/block.hpp]: Yes, but you can use `this->values().sizes()` instead. 

`values_shape()` is here for integration with metatensor (which uses `vector<size_t>`)

File: metatensor-torch/tests/block.cpp [hunk 6]
```diff
@@ -1,4 +1,5 @@
 #include <torch/torch.h>
+#include <iostream>
```
Comment (Luthaf) [inline metatensor-torch/tests/block.cpp:2]:
```suggestion
```

File: metatensor-torch/include/metatensor/torch/block.hpp [hunk 7]
```diff
@@ -170,6 +181,8 @@ class METATENSOR_TORCH_EXPORT TensorBlockHolder: public torch::CustomClassHolder
     /// If this TensorBlock contains gradients, these are gradients w.r.t. this
     /// parameter
     std::string parameter_;
+    /// Shape
+    std::vector<int64_t> shapei64 ;
```
Comment (Luthaf) [inline metatensor-torch/include/metatensor/torch/block.hpp:185]: I don't think we need an extra class variable

**Review by Luthaf — COMMENTED** [review]
This just needs a couple small things

- adding the new function to metatensor/torch/documentation.py
- adding a test in metatensor-torch Python package (to check that register is working fine)
- adding the functions around [here](https://github.com/lab-cosmo/metatensor/blob/ec6ee61ebb6eab326462386171f5cd777f12a3c4/python/metatensor-torch/tests/block.py#L323) to check that the code is compatible with TorchScript

**Review by Luthaf — APPROVED** [review]
I'll merge once CI is happy!

**Pull request merged**